models/Image.js

```diff
@@ -1,76 +1,160 @@
-    image: {
-        type: Types.GcsImage,
-        initial: true,
-        autoCleanup: true,
-        datePrefix: 'YYYYMMDDHHmmss',
-        // TODO move these settings to config
-        bucket: bucket,
-        destination: 'images/',
-        publicRead: true,
-        resize: resizeImage,
-        resizeOpts: [{
-            target: 'desktop',
-            width: 2000,
-            height: null,
-            options: {}
-        }, {
-            target: 'tablet',
-            width: 1200,
-            height: null,
-            options: {}
-        }, {
-            target: 'mobile',
-            width: 800,
-            height: null,
-            options: {}
-        }, {
-            target: 'w400',
-            width: 400,
-            height: null,
-            options: {}
-        }, {
-            target: 'tiny',
-            width: 150,
-            height: null,
-            options: {}
-        }],
-        extractIPTC: extractIPTC
-    },
-    description: {
-        type: String,
-        index: true
-    },
-    copyright: {
-        type: Types.Select,
-        options: 'Creative-Commons, Copyrighted',
-        default: 'Copyrighted', index: true
-    },
-    keywords: {
-        type: String,
-        index: true
-    },
-    sale: {
-        type: Boolean,
-        index: true
+  image: {
+    type: Types.GcsImage,
+    initial: true,
+    autoCleanup: true,
+    datePrefix: 'YYYYMMDDHHmmss',
+    bucket: bucket,
+    destination: '/',
+    publicRead: false,
+    hooks: {
+      postUpload: async (imageData) => {
+        if (typeof imageData !== 'object' || imageData === null) {
+          return imageData;
+        }
+
+        const { filepath, filename, height, width } = imageData;
+
+        if (typeof filename !== 'string' || filename === '') {
+          return imageData;
+        }
+
+        try {
+          await publishResizeTasks(filename);
+        } catch (err) {
+          throw errors.helpers.wrap(err, 'Error', 'Received error while publishing resize events.')
+        }
+
+        try {
+          const iptc = await extractIPTC(filepath);
+          imageData.iptc = iptc;
+        } catch (err) {
+          throw errors.helpers.wrap(err, 'Error', 'Error to extract iptc from images.')
+        }
+
+        const resizedTargets = getResizeMetadata(filename, width, height);
+        imageData.resizedTargets = resizedTargets;
+
+        // Since we set `publicRead` to be false (see config above),
+        // Keystone will only provide authenticated url for those uploaded images.
+        // Therefore, we need to overwrite `imageData.url`, which is used by
+        // @twreporter/keystone/fields/types/gcsimage/GcsImageField.js to display at
+        // `https://keystone-editor.twreporter.org/keystone/images/${imageID}` page,
+        // to be a public one.
+        const mobileImgUrl = _.get(resizedTargets, 'mobile.url');
+        if (mobileImgUrl) {
+          imageData.url = mobileImgUrl;
+        }
```

Comment:
If we always fetch the thumbnail with mobile resolution on keystone, do we need to ensure the availability of image during the uploading process?

Reply:
Currently, in our new design, it's not easy to ensure the availability.

The new logic is:
If there is something wrong at publishing resize tasks, we think the uploading fails.
If publishing resize tasks does well, we think the uploading succeeds.

The resize tasks are async processing, I don't have a clue to ensure the availability.

Reply:
> Currently, in our new design, it's not easy to ensure the availability.
> 
> The new logic is:
> If there is something wrong at publishing resize tasks, we think the uploading fails.
> If publishing resize tasks does well, we think the uploading succeeds.
> 
> The resize tasks are async processing, I don't have a clue to ensure the availability for now.

Yes, it's not easy to monitor for the status of the whole resized tasks. Current problem would be the list page probably displays the `not found image` if the image is not available. Would it be better to poll the `mobile` image after `postUpload` to ensure the availability?

Reply:
> Would it be better to poll the mobile image after postUpload to ensure the availability?

If we poll the just uploaded photo here, the uploading process will be stuck here.
In addition, we don't have the mechanism to know the resize status, so we could not have a good way to know when to stop polling.

Since this task (https://kanbanflow.com/t/g8YQwwJ7) will show messages to uploaders, could we render `影像處理中，請稍候。` or `影像處理中，請於 slack channel 確認影像狀態。` ... etc, if mobile URL is not ready?

Reply:
> > Would it be better to poll the mobile image after postUpload to ensure the availability?
> 

The goal here is to ensure that the photography department won't feel confused about the uploading after the new system deployment.

One thing, I would like to confirm first was that: `Does the photographers see the image on list page as soon as the upload process is done?`

> If we poll the just uploaded photo here, the uploading process will be stuck here.
> In addition, we don't have the mechanism to know the resize status, so we could not have a good way to know when to stop polling.

If the image should be listed immediately,
could we simplify the status to `object exist or not` to make this possible?

My attempt described in below pseudo codes.

```
status_code =  Get(Image_url)

while(status_code != 200)
{
sleep(10ms)
// try again
Get(image_url)
}
``` 

> 
> Since this task (https://kanbanflow.com/t/g8YQwwJ7) will show messages to uploaders, could we render `影像處理中，請稍候。` or `影像處理中，請於 slack channel 確認影像狀態。` ... etc, if mobile URL is not ready?

Yes, that would be helpful, but this would require the photographer to always refresh the page after the system deploys. I'm trying to mitigate the inconvenience.

Reply:
> If the image should be listed immediately,
> could we simplify the status to `object exist or not` to make this possible?
> 
> My attempt described in below pseudo codes.
> 
> ```
> status_code =  Get(Image_url)
> 
> while(status_code != 200)
> {
> sleep(10ms)
> // try again
> Get(image_url)
> }
> ```
> 

I have two concerns about this approach:

1. The uploading process will only be finished if `hooks.postUpload` function has done. Therefore, if we poll the mobile image url here, end users can do nothing only if polling is done.

2. For the pseudo codes, what if resize task fails? 
The while loop could become an infinite loop.
At the end, we have to set up the timeout config, but what value will be appropriate?
If we set up the timeout, we still have to show message to end users when program timeouts.

Reply:
To approach the issue further, we need to make sure the use case.

As I stated above, we need to confirm: `Does the photographers see the image on list page as soon as the upload process is done?`. Could you confirm that?

> I have two concerns about this approach:
> 
> 1. The uploading process will only be finished if `hooks.postUpload` function has done. Therefore, if we poll the mobile image url here, end users can do nothing only if polling is done.

How is the effort to render a loading text or image here?

> 2. For the pseudo codes, what if resize task fails?
>    The while loop could become an infinite loop.
>    At the end, we have to set up the timeout config, but what value will be appropriate?
>    If we set up the timeout, we still have to show message to end users when program timeouts.

The timeout could be adhere to the cloud-function timeout.

However, we should confirm the use case first and push the discussion further if required.